### PR TITLE
Add session management for wallet connect

### PR DESCRIPTION
**Pull request by cammellos**
This PR adds support for managing wallet connect sessions.

Currently changing of account is not supported, that will be implemented separately.

What's implemented is:

- Storing sessions (login/logout) should work as expected
- Disconnecting a session from mobile 
- Disconnecting a session from the website while the app is online

I think the logic could be further simplified, there's some really odd bits that I had to introduce (storing `peerId` onto the connector object), but could not find a better way. The only other reference (rainbow wallet), disconnects all the sessions for that URL.

status: ready

### CODE REVIEW

**Comment by briansztamfater** [comment]
Does this PR add session management for Wallet Connect v2 also or only for v1?

**Comment by cammellos** [comment]
> Does this PR add session management for Wallet Connect v2 also or only for v1?

Only 1 for now, forgot to specify

File: src/status_im/ui/screens/wallet/manage_connections/styles.cljs [hunk 1]
```diff
@@ -0,0 +1,43 @@
+(ns status-im.ui.screens.wallet.manage-connections.styles
+  (:require [quo.design-system.colors :as colors]))
+
+(def daap-icon
```
Comment (Parveshdhull) [inline src/status_im/ui/screens/wallet/manage_connections/styles.cljs]: dapp?

File: src/status_im/ui/screens/wallet/manage_connections/views.cljs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+(defn print-session-info [session]
+  (let [peer-meta (get-in session [:params 0 :peerMeta])
+        peer-id (get-in session [:params 0 :peerId])
+        name (get peer-meta :name)
+        url (get peer-meta :url)
+        account (get-in session [:params 0  :accounts 0])
+        icons (get peer-meta :icons)
+        filtered-icons (status-im.utils.utils/exclude-svg-resources icons)
+        icon-uri (when (and filtered-icons (pos? (count filtered-icons))) (first filtered-icons))
```
Comment (Parveshdhull) [inline src/status_im/ui/screens/wallet/manage_connections/views.cljs]: `(when (seq filtered-icons) (first filtered-icons))` ?

File: src/status_im/utils/utils.cljs [hunk 3]
```diff
@@ -131,3 +131,9 @@
 (defn safe-replace [s m r]
   (when (string? s)
     (string/replace s m r)))
+
+(defn is-svg? [some-string]
+  (not (string/includes? some-string ".svg")))
```
Comment (Parveshdhull) [inline src/status_im/utils/utils.cljs]: isn't `is-svg?` should be true when it includes ".svg"? Or function should be renamed to `is-not-svg?`

File: src/status_im/ui/screens/wallet/manage_connections/views.cljs (hunk 2 above)
Comment (Parveshdhull) [reply src/status_im/ui/screens/wallet/manage_connections/views.cljs]: Or just plain, `(first filtered-icons)` as we don't need to check condition at all, it still be nil when condition will fail.

File: src/status_im/utils/utils.cljs (hunk 3 above)
Comment (siddarthkay) [reply src/status_im/utils/utils.cljs]: Hmm you're right.. 
The reason I made this was to filter out svgs from the array.
probably `filter-svg` would be a better rename.

File: src/status_im/ui/screens/wallet/manage_connections/styles.cljs (hunk 1 above)
Comment (siddarthkay) [reply src/status_im/ui/screens/wallet/manage_connections/styles.cljs]: the naming was dApp actually : Decentralised App

Comment (siddarthkay) [reply src/status_im/ui/screens/wallet/manage_connections/styles.cljs]: I will rename :)

**Review by Parveshdhull — APPROVED** [review]
please also replace other instances of `(when (and filtered-icons ...`

File: src/status_im/utils/utils.cljs [hunk 4]
```diff
@@ -131,3 +131,9 @@
 (defn safe-replace [s m r]
   (when (string? s)
     (string/replace s m r)))
+
+(defn is-not-svg? [some-string]
+  (not (string/includes? some-string ".svg")))
+
+(defn exclude-svg-resources [lst]
+  (filter is-not-svg? lst))
```
Comment (rasom) [inline src/status_im/utils/utils.cljs]: **not** like it actually matters, but less parenthesis is less parenthesis 
```clojure
(defn is-svg? [some-string]
  (string/includes? some-string ".svg"))
 
(defn exclude-svg-resources [lst]
  (remove is-svg? lst))
```

Comment (rasom) [reply src/status_im/utils/utils.cljs]: and then probably just `svg?` https://github.com/bbatsov/clojure-style-guide#naming-predicates

Comment (rasom) [reply src/status_im/utils/utils.cljs]: and likely not a big deal, but shouldn't it be `(string/ends-with? some-string ".svg")`? I should probably stop looking at those lines

**Comment by pavloburykh** [comment]
### ISSUE 1 Transaction signing pop does not appear while walletconnect connected on Mainnet

In case of performing initial walletconnect connection on mainnet - I don't get a promt  in Status app to sign transaction triggered on uniswap side.

**Preconditions:** Mainnet is selected in Status app, Wallet connect is turned on

**Steps:**
1.  Open https://app.uniswap.org/chain=mainnet#/swap?chain=mainnet
2. Perform connection via walletconnect by scanning QR code
3. Initiate any swap operation in Uniswap app
4. See if signing pop up appears in Status app

**Actual result:** signing pop up does not appear, unable to sign transaction

**Expected result:** signing pop up appears. It is possible to sign transaction. 


https://user-images.githubusercontent.com/97245802/184630586-1fb1082b-c368-4750-a0ea-82601eddd3c8.mp4

**Comment by pavloburykh** [comment]
### ISSUE 1 Transaction can be signed in case of different networks selected on Status side and connected via walletconnect app

**Steps:**
1.  Open https://app.uniswap.org/chain=mainnet#/swap?chain=mainnet
2. Perform connection via walletconnect by scanning QR code
3. Initiate any swap operation in Uniswap app
4. See if signing pop up appears in Status app

**Actual result:** signing pop up appears in Status app

**Expected result:** signing pop up should not appear. User should not be allowed to sign transactions while being on different connected app network, where transaction is initiated. 

Video from Status app

https://user-images.githubusercontent.com/97245802/184632892-1e9eb05d-ff88-47fd-9b86-7e0c5dd79ff8.mp4

Video from opened on destop browser uniswap app

https://user-images.githubusercontent.com/97245802/184632961-99566dcb-1fe1-472e-888b-e8081acf6dfc.mp4

**Review by briansztamfater — APPROVED** [review]

**Comment by pavloburykh** [comment]
#### ISSUE 2 The transaction signing popup does not appear if you cancel the transaction signature once, with pressing `Show again signing phrase` before that, and initiate the transaction signature again (on Mainnet but if it was connected to Goerli before)

**Preconditions:** Goerli is selected in Status app, Wallet connect is turned on

**Steps:**
1.  Open https://app.uniswap.org/chain=mainnet#/swap?chain=mainnet
2.  Perform connection via walletconnect by scanning QR code 
3. Change Status network to Mainnet
4. Initiate Swap operation in uniswap
5. Tap `Show again signing phrase` and cancel signing up in Status app (`Show again signing phrase` appears again here)
6. Initiate another swap operation in Uniswap app
7. See if signing pop up appears in Status app

**Actual result:** signing pop up does not appear, unable to sign transaction. After relogin in Status signing works fine

**Expected result:** signing pop up appears. It is possible to sign transaction. 



https://user-images.githubusercontent.com/67952253/184845883-c0645b13-57bf-438c-9d9c-95bdee7fa79f.mp4




### ISSUE 3 Wallet is displayed connected in app connection settings if it was disconnected in uniswap while Status app being offline

Steps:   1. Connect Status app to uniswap via wallet connect
2. Logout Status app
3. Disconnect wallet in uniswap
4. Login Status app
5. Go to wallet tab - open App connections settings
6. See if uniswap app is displayed among connections

Actual result: uniswap app is still displayed in connections


### ISSUE 4 Two operations in a row appear for signing in Status app on Mainnet instead of one initiated in dapp (if Status was on Goerli before switching to Mainnet, please follow steps below)

**Preconditions:** Goerli is selected in Status app, Wallet connect is turned on

**Steps:**
1.  Open https://app.uniswap.org/chain=mainnet#/swap?chain=mainnet
2.  Perform connection via walletconnect by scanning QR code 
3. Change Status network to Mainnet
4. Initiate Swap operation in uniswap
5. Cancel or sign transaction in Status
6. See if another signing pop up appears after signing/canceling first operation

**Actual result:** signing pop up appears for the second time

**Expected result:** signing pop up appears only once for one initiated transaction

**Comment by qoqobolo** [comment]
@cammellos 

### ISSUE 5/Q: Should it be possible to change the wallet account in the `Manage app connections` settings?

The account button looks active and clickable, but nothing happens on pressing it. 
If it should not, then it would probably be better to make this button inactive to avoid ambiguity.

https://user-images.githubusercontent.com/67952253/184847943-992fd344-1146-4bda-89d5-6ecdc708be77.mp4

### ISSUE 6: Theme doesn't apply to a dapp name in the `Manage app connections` settings when changing it on the device or in the app settings. 

**Steps:**
1. Connect to Uniswap
2. Change the mode
3. Open `Manage app connections`

https://user-images.githubusercontent.com/67952253/184849104-26bcd986-169e-4b5d-8f56-f69b361761ae.mp4

**Comment by qoqobolo** [comment]
@cammellos 

### ISSUE 7: Connections list is not aligned on its right side

Screenshot 2022-08-16 at 14 55 26

### ISSUE/QUESTION 8: What is the expected behavior for the case when we connect to the same DApp from one mobile device but to two different desktop clients?

Not sure this is a relevant case, but wondering if it should detect that this device is already connected to this DApp or not (If at all possible).
Currently the same DApp connection is displayed twice in the list.

**Steps:**
1. Open the Uniswap QR on one desktop device > Connect your mobile device with WalletConnect
2. Open the Uniswap QR on another device > Connect your mobile device (same wallet) with WalletConnect
3. Check connections in `Manage app connections`

https://user-images.githubusercontent.com/67952253/184886556-9da4d173-2927-4b69-9d5e-8faab674e27f.mp4

### ISSUE 9: Cannot disconnect from the Zapper DApp 

**Steps:**
1. Switch to Mainnet, enable WalletConnect
2. Connect to https://zapper.fi/
3. Go to `Manage app connections` > Remove connection
4. Check the desktop Zapper DApp 

**Expected result:** wallet is disconnected
**Actual result:** wallet is still connected even after relogin

Other DApps from this list https://explorer.walletconnect.com/ work fine with these steps (can disconnect my wallet by pressing `Remove` on my mobile device), so it's probably an issue on the DApp side. WDYT? 


https://user-images.githubusercontent.com/67952253/184889744-249db4be-dfde-425b-b1d4-6d13dbd98a96.mp4

**Comment by cammellos** [comment]
@qoqobolo 

> ISSUE/QUESTION 8: What is the expected behavior for the case when we connect to the same DApp from one mobile device but to two different desktop clients?

For now, I'd say that's the expected behavior, some clients make so that only one connection is available, but for now I'd keep multiple, just in case people have multiple desktops/etc

> ISSUE 9: Cannot disconnect from the Zapper DApp

It's most likely a problem on their side, but I'll take a look

> ISSUE 5/Q: Should it be possible to change the wallet account in the Manage app connections settings?

I think we will handle this in a separate PR, what do you think?

I am taking a look at the rest of the issues in the meantime, thanks for testing!

**Comment by cammellos** [comment]
@qoqobolo 

Issue 2 I could not replicate, I think it should be fixed

Issue 3 we can't prevent, I have checked similar wallets (raindbow), and it's the same behavior, the library unfortunately does not notifies us if there was a disconnect while offline.

Issue 4 is fixed

Issue 5 will be implemented separately, but won't make it for 1.20

Issue 6,7 should be fixed

After we can decide whether to keep under a toggle for 1.20, I'd be inclined to remove the toggle and have it on for all users, but let me know what you think.

Thanks for testing!

**Comment by qoqobolo** [comment]
@cammellos thanks a lot for the fixes and answers!

> ISSUE 5/Q: Should it be possible to change the wallet account in the Manage app connections settings?
I think we will handle this in a separate PR, what do you think?

Sure, I'll register it separately.

> Issue 2 I could not replicate, I think it should be fixed

Hmm, I still can reproduce it. Just to make sure: do you tap `Show again signing phrase` when this pop-up appears for the first time (it's not reproducible if you tap `Okay, got it`)? If after that you cancel transaction signing, the pop-up appears a second time. And after that, it won't be possible to call the signature dialog for the second time (until relogin).

https://user-images.githubusercontent.com/67952253/185159502-15591776-2ed2-4581-9df7-442b968b9c4c.mp4

> Issue 4 is fixed

And I still can also reproduce Issue 4 with the same steps: the signing dialog appears twice.
On the video below I sign a transaction on Goerli, but I switched from Goerli to Mainnet and then to Goerli again.
On Mainnet it's reproducible as well. 

https://user-images.githubusercontent.com/67952253/185169016-fa1cb9e7-23b1-4ff3-a1f7-89fcfb47784f.mp4

> After we can decide whether to keep under a toggle for 1.20, I'd be inclined to remove the toggle and have it on for all users, but let me know what you think.

I think we will be absolutely okay with removing the toggle! Let's just see how the full e2e suit passes after the final fixes and decide for sure.

**Comment by qoqobolo** [comment]
Confirm that issues 2 and 4 are fixed, thanks @cammellos!
Launching full e2e suit and making the final check.
Thank you for your wait and patience.

**Comment by pavloburykh** [comment]
> Confirm that issues 2 and 4 are fixed, thanks @cammellos! Launching full e2e suit and making the final check. Thank you for your wait and patience.

looks good to me too. Waiting for removing wallet connect toggle and will be ready for merge.

**Comment by qoqobolo** [comment]
@cammellos thanks a lot for the fixes! But unfortunately, I see now two more issues :( (they could have the same reason though)

### ISSUE 10: Balance is displayed incorrectly in the desktop DApp

There is no amount either on Goerli or on Mainnet. 

Screenshot 2022-08-19 at 10 21 39

### ISSUE 11: The DApp doesn't see the ENS name

**Steps:**
1. Set up ENS in the profile
2. Connect wallet to Uniswap

Tried on Mainnet.
I can see that Pavlo's name and balance were displayed before
Screenshot 2022-08-19 at 10 24 01

**Comment by qoqobolo** [comment]
@cammellos thank you for your awesome work and patience!
PR is tested and can be merged.

**Comment by cammellos** [comment]
@qoqobolo @pavloburykh thanks a lot for your patience and testing!

**Pull request merged**